src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala

```diff
@@ -0,0 +1,228 @@
+      scheduleNextWorkflowQuery
+  }
+
+  def scheduleNextWorkflowQuery: Cancellable = {
+    system.scheduler.scheduleOnce(pollInterval, self, LookForWorkflows)
+  }
+}
+
+trait WorkflowSubmission extends FutureSupport with LazyLogging with MethodWiths {
+  val dataSource: SlickDataSource
+  val methodRepoDAO: MethodRepoDAO
+  val googleServicesDAO: GoogleServicesDAO
+  val executionServiceDAO: ExecutionServiceDAO
+  val batchSize: Int
+  val credential: Credential
+  val pollInterval: FiniteDuration
+
+  import dataSource.dataAccess.driver.api._
+
+  //Get a blob of unlaunched workflows, flip their status, and queue them for submission.
+  def getUnlaunchedWorkflowBatch()(implicit executionContext: ExecutionContext): Future[WorkflowSubmissionMessage] = {
+    val unlaunchedWfOptF = dataSource.inTransaction { dataAccess =>
+      //grab a bunch of unsubmitted workflows
+      dataAccess.workflowQuery.findUnsubmittedWorkflows().take(batchSize).result map { wfRecs =>
+        if (wfRecs.nonEmpty) wfRecs
+        else {
+          //they should also all have the same submission ID
+          val wfsWithASingleSubmission = wfRecs.filter(_.submissionId == wfRecs.head.submissionId)
+          //instead: on update, if we don't get the right number back, roll back the txn
+          //TODO: biden has an incoming change here, with optimistic locking
+          dataAccess.workflowQuery.batchUpdateStatus(wfsWithASingleSubmission.map(_.id), WorkflowStatuses.Launching)
+          wfsWithASingleSubmission
+        }
+      }
+    }
+
+    //if we find any, next step is to submit them. otherwise, look again.
+    unlaunchedWfOptF.map {
+      case workflowRecs if workflowRecs.nonEmpty => SubmitWorkflowBatch(workflowRecs.map(_.id))
+      case _ => LookForWorkflows
+    }
+  }
+
+  def buildWorkflowOpts(workspace: WorkspaceRecord, submissionId: UUID, user: RawlsUser, token: String, billingProject: RawlsBillingProject) = {
+    ExecutionServiceWorkflowOptions(
+      s"gs://${workspace.bucketName}/${submissionId}",
+      workspace.namespace,
+      user.userEmail.value,
+      token,
+      billingProject.cromwellAuthBucketUrl
+    )
+  }
+
+  def getWdl(methodConfig: MethodConfiguration): ReadWriteAction[String] = {
+    withMethod(methodConfig.methodRepoMethod.methodNamespace, methodConfig.methodRepoMethod.methodName, methodConfig.methodRepoMethod.methodVersion, getUserInfo(credential)) { method =>
+      withWdl(method) { wdl =>
+        DBIO.successful(wdl)
+      }
+    }
+  }
+
+  //submit the batch of workflows with the given ids
+  def submitWorkflowBatch(workflowIds: Seq[Long]): Future[WorkflowSubmissionMessage] = {
+
+    val workflowBatchFuture = dataSource.inTransaction { dataAccess =>
+      for {
+      //Load a bunch of things we'll need to reconstruct information:
+      //The list of workflows in this submission
+        wfRecs <- dataAccess.workflowQuery.findWorkflowByIds(workflowIds).result
+        submissionId = wfRecs.head.submissionId
```

Comment:
Turns out that Scala doesn't like assignments in for comprehensions: the semantics [are weird and mindbendy](http://stackoverflow.com/questions/29433727/the-for-comprehension-with-assignment-loops-in-wrong-order).